src/main/kotlin/com/nylas/NylasClient.kt

```diff
@@ -0,0 +1,248 @@
+      throwAndCloseOnFailedRequest(response)
+      val responseBody = response.body()
+      if (resultType == null || responseBody == null) {
+        throw Exception("Unexpected null response body")
+      }
+
+      val adapter = JsonHelper.moshi().adapter<T>(resultType)
+
+      return adapter?.fromJson(responseBody.source()) ?: throw Exception("Failed to deserialize response body")
+    }
+  }
+
+  @Throws(IOException::class, NylasApiError::class)
+  private fun throwAndCloseOnFailedRequest(response: Response) {
+    if (!response.isSuccessful) {
+      val responseBody = response.body()!!.string()
+      response.close()
+      val parsedError = JsonHelper.moshi().adapter(NylasApiErrorResponse::class.java)
+        .fromJson(responseBody)
+      if (parsedError?.error != null) {
+        throw parsedError.error
+      } else {
+        throw NylasApiError("unknown", "Unknown error received from the API: $responseBody")
+      }
+    }
+  }
+
+  private fun buildUrl(path: String, queryParams: IQueryParams?): HttpUrl.Builder {
+    var url = newUrlBuilder().addPathSegments(path)
+    if (queryParams != null) {
+      url = addQueryParams(url, queryParams.convertToMap())
+    }
+    return url
+  }
+
+  private fun addQueryParams(url: HttpUrl.Builder, params: Map<String, Any>): HttpUrl.Builder {
+    for ((key, value) in params) {
+      url.addQueryParameter(key, value.toString())
+    }
+    return url
+  }
+
+  /**
+   * A NylasClient.Builder allows applications to customize the Nylas http access
+   * by choosing a different base url or modifying http client options.
+   */
+  data class Builder(
+    private val apiKey: String,
+  ) {
+    // TODO::is this good?
+    private var baseUrl: String = DEFAULT_BASE_URL
+    private var httpClient: OkHttpClient.Builder = defaultHttpClient()
+    fun baseUrl(baseUrl: String) = apply { this.baseUrl = baseUrl }
+
+    /**
+     * Access the OkHttpClient.Builder to allow customization.
+     * By default, the NylasClient configures it as follows:
+     * .protocols(Arrays.asList(Protocol.HTTP_1_1))
+     * .connectTimeout(60, TimeUnit.SECONDS)
+     * .readTimeout(60, TimeUnit.SECONDS)
+     * .writeTimeout(60,  TimeUnit.SECONDS)
+     * .addNetworkInterceptor(new HttpLoggingInterceptor()
+     */
+    fun httpClient(httpClient: OkHttpClient.Builder) = apply { this.httpClient = httpClient }
+
+    fun build() = NylasClient(apiKey, httpClient, baseUrl)
+  }
+
+  companion object {
+    const val DEFAULT_BASE_URL = "https://api-staging.us.nylas.com/"
+    const val EU_BASE_URL = "https://ireland.api.nylas.com/"
+    private fun defaultHttpClient(): OkHttpClient.Builder {
+      return OkHttpClient.Builder()
+        .connectTimeout(60, TimeUnit.SECONDS)
+        .readTimeout(60, TimeUnit.SECONDS)
+        .writeTimeout(60, TimeUnit.SECONDS)
+        .protocols(listOf(Protocol.HTTP_1_1))
+//                .addNetworkInterceptor(HttpLoggingInterceptor())
+    }
+  }
```

Comment:
@mrashed-dev do we access this like static properties?

Reply:
@AaronDDM thanks for calling this out. This is not intended to be used by the end user, here we're stashing default values. I have updated the `Regions` enum to contain the different API urls and that's what the user can use to set a different URL:

```java
NylasClient nylas = new NylasClient.Builder("API_KEY")
  .baseUrl(Region.EU.getNylasApiUrl())
  .build();
```